projects/gloo/pkg/plugins/tunneling/plugin.go

```diff
@@ -40,136 +47,175 @@ func (p *plugin) Name() string {
+	if httpProxyHostname == "" {
+		return nil, nil, nil, nil
+	}
+
+	clusterName := translator.UpstreamToClusterName(us.GetMetadata().Ref())
+
+	// find the cluster to update
+	cluster := findClusters(inClusters, clusterName)
+	if cluster == nil {
+		return nil, nil, nil, fmt.Errorf("cluster %s not found", clusterName)
+	}
+
+	// change the original cluster name to avoid conflicts with the new cluster
+	newOriginalClusterName := clusterName + OriginalClusterSuffix
+
+	// use an in-memory pipe to ourselves (only works on linux)
+	forwardingPipe := "@/" + clusterName
+
+	var originalTransportSocket *envoy_config_core_v3.TransportSocket
+	tunnelingHeaders := envoyHeadersFromHttpConnectHeaders(us)
+
+	// create the listener with the tunneling configuration and point it the original clusters
+	newListener, err := generateForwardingTcpListener(clusterName, newOriginalClusterName,
+		forwardingPipe, httpProxyHostname, tunnelingHeaders)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	// we copy the transport socket to the generated cluster.
+	// the generated cluster will use upstream TLS context to leverage TLS origination;
+	if cluster.GetTransportSocket() != nil {
+		tmp := *cluster.GetTransportSocket()
+		originalTransportSocket = &tmp
+	}
+
+	// update the original cluster to use the new SSL configuration for the HTTP proxy, if provided
+	if us.GetHttpConnectSslConfig() != nil {
+		// user told us to configure ssl for the http connect proxy
+		cfg, err := utils.NewSslConfigTranslator().ResolveUpstreamSslConfig(params.Snapshot.Secrets,
+			us.GetHttpConnectSslConfig())
+		if err != nil {
+			// if we are configured to warn on missing tls secret and we match that error, add a
+			// warning instead of error to the report.
+			if params.Settings.GetGateway().GetValidation().GetWarnMissingTlsSecret().GetValue() &&
+				errors.Is(err, utils.SslSecretNotFoundError) {
+				warningList = append(warningList, &translator.Warning{
+					Message: err.Error(),
+				})
+			} else {
+				return nil, nil, nil, err
+			}
+		} else {
+			typedConfig, err := utils.MessageToAny(cfg)
+			if err != nil {
+				return nil, nil, nil, err
+			} else {
+				cluster.TransportSocket = &envoy_config_core_v3.TransportSocket{
+					Name:       wellknown.TransportSocketTls,
+					ConfigType: &envoy_config_core_v3.TransportSocket_TypedConfig{TypedConfig: typedConfig},
 				}
 			}
 		}
+	} else {
+		cluster.TransportSocket = nil
 	}
 
-	return generatedClusters, nil, nil, generatedListeners, nil
+	// when we encapsulate in HTTP Connect the tcp data being proxied will
+	// be encrypted (thus we don't need the original transport socket metadata here)
+	cluster.TransportSocketMatches = nil
+
+	// generate new cluster with original cluster's name and transport socket that points to
+	// the new listener's pipe
+	newCluster := generateForwardingCluster(clusterName, forwardingPipe, originalTransportSocket)
+
+	// update the original route's cluster name
```

Comment:
nit: This feels like a critical change to support this feature. I wonder if it's worht it to call this comment out more loudly to clarify what the new cluster name is

Reply:
Ok. For clarity, you want the comment it's self to be more emphatic or it called out with more emphasis somewhere else?

Reply:
More details have been added to the comment.